rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+    - All result observables are completed
+   
+- on every emission of the merged observable, the results will be aggregated to the existing list and sorted following 
+the logic defined in the [results aggregation](#results-aggregation) section
+
+A very naive implementation of this behavior would be:
+
+```ts
+search(
+  term: string,
+  options: GlobalSearchOptions,
+  request: KibanaRequest
+): Observable<GlobalSearchResponse> {
+  const fromProviders$ = this.providers.map(p =>
+    p.find(term, options, contextFromRequest(request))
+  );
+  const results: GlobalSearchResult[] = [];
+  return merge([...fromProviders$]).pipe(
+    takeUntil(timeout$),
+    map(newResults => {
+      results.push(...newResults);
+      return order(results);
+    }),
+  );
+}
+```
+
+### searching from the client side
+
+When calling `GlobalSearchServiceStart.find` from the public-side service:
+
+- The service will call:
+  - the server-side API via an http call to fetch results from the server-side result providers
+  - `find` on each client-side registered result provider and collect the resulting observables
+
+- Then, the service will merge every result observable and trigger the next step on every emission until either
+  - A predefined timeout duration is reached
+  - All result observables are completed
+   
+- On every emission of the merged observable, the results will be aggregated to the existing list and sorted following 
+the logic defined in the [results aggregation](#results-aggregation) section
+
+A very naive implementation of this behavior would be:
+
+```
+search(
+  term: string,
+  options: GlobalSearchOptions,
+): Observable<GlobalSearchResponse> {
+  const fromProviders$ = this.providers.map(p =>
+    p.find(term, options)
+  );
+  const fromServer$ = of(this.fetchServerResults(term, options))
+
+  const results: GlobalSearchResult[] = [];
+  return merge([...fromProviders$, fromServer$]).pipe(
+    takeUntil(timeout$),
+    map(newResults => {
+      results.push(...newResults);
+      return order(results);
+    }),
+  );
+}
+```
+
+Notes: 
+- Due to the complexity of the process, the initial implementation will not be streaming results from the server,
+meaning that all results from server-side registered providers will all be fetched at the same time (via a 'classic' 
+http call to the GS endpoint). The observable-based API architecture is ready for this however, and the enhancement 
+could be added at a later time.
+
+### results aggregation
+
+On every emission of an underlying provider, the service will aggregate the new results with the existing list and 
+sort the results following this logic before emitting them:
+
+- Results will be sorted by ascending `type` ordinal value.
+- Results of a same `type` will then be sorted by descending `score` value.
+
+This is an equivalent of the following lodash call:
+
+```ts
+const sorted = _.sortBy(unsorted, [r => SEARCH_TYPE[r.type], 'score'], ['asc', 'desc']);
+```
```

Comment:
@elastic/kibana-core-ui this is what was discussed in https://github.com/elastic/kibana/issues/61657. Please confirm this is still alright.

Reply:
Global search, within Kibana, is considered a key component to navigation. That being the case, we desire to have applications always appear first, before all other object types. Whether that is accomplished through sorting or some sort of weighted result, that is our desired end result.

Reply:
> we desire to have applications always appear first, before all other object types

Yea, this is where I'm mixed. I'm not sure to which extend this logic belongs to `GS` API or more to the `searchbar` implementation

Maybe we should just sort by score, and let `searchbar` reorder it following its own logic (moving a given type of results on top from an already sorted list is trivial)

@joshdover WDYT?

Reply:
That would make sense because of https://github.com/elastic/kibana/pull/64284#discussion_r414581051